File: src/lib/kafe/src/main/java/kafe/BaseShould.java
package kafe;

import java.util.function.Function;

/**
 * BaseShould extends BaseAction for TDT pattern.
 * Processes each row in table-driven testing.
 */
public abstract class BaseShould<R> extends BaseAction<Object, R> {
    public BaseShould(String name, Function<R, R> shouldCb) {
        super(name, shouldCb);
    }
    
    // Alias performAction to shouldDo for TDT pattern
    @Override
    public abstract R performAction(
        R store,
        Function<R, R> actionCB,
        ITTestResourceConfiguration testResourceConfiguration,
        Object artifactory
    );
    
    // Alias test to shouldTest for TDT pattern
    public R shouldTest(R store, ITTestResourceConfiguration testResourceConfiguration) throws Exception {
        return super.test(store, testResourceConfiguration);
    }
}

File: src/lib/golingvu/base_should.go
package golingvu

// BaseShould extends BaseAction for TDT pattern
type BaseShould struct {
	*BaseAction
	CurrentRow []interface{}
	RowIndex   int
}

// NewBaseShould creates a new BaseShould instance
func NewBaseShould(name string, shouldCB interface{}) *BaseShould {
	baseAction := NewBaseAction(name, shouldCB)
	return &BaseShould{
		BaseAction: baseAction,
		RowIndex:   -1,
		CurrentRow: []interface{}{},
	}
}

// PerformAction implements BaseAction's abstract method
func (bs *BaseShould) PerformAction(
	store interface{},
	actionCB interface{},
	testResource interface{},
	artifactory func(string, interface{}),
) (interface{}, error) {
	// This should be implemented by the adapter
	return store, nil
}

// SetRowData sets current row data
func (bs *BaseShould) SetRowData(rowIndex int, rowData []interface{}) {
	bs.RowIndex = rowIndex
	bs.CurrentRow = rowData
}

// ProcessRow processes the current row
func (bs *BaseShould) ProcessRow(
	store interface{},
	testResourceConfiguration ITTestResourceConfiguration,
	artifactory func(string, interface{}),
) (interface{}, error) {
	return bs.Test(store, testResourceConfiguration, artifactory)
}
